src/Compilers/CSharp/Test/Emit/CodeGen/ForeachTest.cs

```diff
@@ -302,6 +303,417 @@ .maxstack  4
+  IL_0049:  call       ""void System.Console.Write(int)""
+  IL_004e:  ret
+}");
+        }
+
+        [Fact]
+        public void TestReadOnlySpan()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System;
+
+class Test
+{
+    public static void Main()
+    {       
+        var sp = new ReadOnlySpan<int>(new[] {1, 2, 3});
+        foreach(var i in sp)
+        {
+            Console.Write(i);
+        }
+    }
+}
+
+", TestOptions.ReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: "123").VerifyIL("Test.Main", @"
+{
+  // Code size       56 (0x38)
+  .maxstack  3
+  .locals init (System.ReadOnlySpan<int> V_0,
+                int V_1)
+  IL_0000:  ldc.i4.3
+  IL_0001:  newarr     ""int""
+  IL_0006:  dup
+  IL_0007:  ldtoken    ""<PrivateImplementationDetails>.__StaticArrayInitTypeSize=12 <PrivateImplementationDetails>.E429CCA3F703A39CC5954A6572FEC9086135B34E""
+  IL_000c:  call       ""void System.Runtime.CompilerServices.RuntimeHelpers.InitializeArray(System.Array, System.RuntimeFieldHandle)""
+  IL_0011:  newobj     ""System.ReadOnlySpan<int>..ctor(int[])""
+  IL_0016:  stloc.0
+  IL_0017:  ldc.i4.0
+  IL_0018:  stloc.1
+  IL_0019:  br.s       IL_002d
+  IL_001b:  ldloca.s   V_0
+  IL_001d:  ldloc.1
+  IL_001e:  call       ""ref readonly int System.ReadOnlySpan<int>.this[int].get""
+  IL_0023:  ldind.i4
+  IL_0024:  call       ""void System.Console.Write(int)""
+  IL_0029:  ldloc.1
+  IL_002a:  ldc.i4.1
+  IL_002b:  add
+  IL_002c:  stloc.1
+  IL_002d:  ldloc.1
+  IL_002e:  ldloca.s   V_0
+  IL_0030:  call       ""int System.ReadOnlySpan<int>.Length.get""
+  IL_0035:  blt.s      IL_001b
+  IL_0037:  ret
+}");
+        }
+
+        [Fact]
+        public void TestSpanNoIndexer()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System;
+
+class Test
+{
+    public static void Main()
+    {       
+        var sp = new Span<int>(new[] {1, 2, 3});
+        foreach(var i in sp)
+        {
+            Console.Write(i);
+        }
+    }
+}
+
+", TestOptions.ReleaseExe);
+
+            comp.MakeMemberMissing(WellKnownMember.System_Span_T__get_Item);
+
+            CompileAndVerify(comp, expectedOutput: "123").VerifyIL("Test.Main", @"
+{
+  // Code size       57 (0x39)
+  .maxstack  4
+  .locals init (System.Span<int> V_0, //sp
+                System.Span<int>.Enumerator V_1)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  ldc.i4.3
+  IL_0003:  newarr     ""int""
+  IL_0008:  dup
+  IL_0009:  ldtoken    ""<PrivateImplementationDetails>.__StaticArrayInitTypeSize=12 <PrivateImplementationDetails>.E429CCA3F703A39CC5954A6572FEC9086135B34E""
+  IL_000e:  call       ""void System.Runtime.CompilerServices.RuntimeHelpers.InitializeArray(System.Array, System.RuntimeFieldHandle)""
+  IL_0013:  call       ""System.Span<int>..ctor(int[])""
+  IL_0018:  ldloca.s   V_0
+  IL_001a:  call       ""System.Span<int>.Enumerator System.Span<int>.GetEnumerator()""
```

Comment:
I find this surprising. Span is a well-known type. Why not just report a diagnostic?

Reply:
Emitting as a `for` is an optimization. We don't have to do it if we can't.

---
In reply to: [153605695](https://github.com/dotnet/roslyn/pull/23363#discussion_r153605695) [](ancestors = 153605695)